Add flag to derive JobSpec from local changes

bbguimaraes opened the pull request:
/cc @droslean @petr-muller @stevekuznetsov

To be used in concert with the staging namespace for cases where we need to change more than the ProwJob. Only the manner in which we derive `JobSpec` is changed, so it should automatically pick up new additions (e.g. template tests) to the CI job.

Includes a few minor changes that didn't seem appropriate as separate PRs.

bbguimaraes committed:
Generalize git function

bbguimaraes committed:
Set git user and email in integration tests

droslean commented on cmd/pj-rehearse/main.go:
```diff
@@ -142,6 +178,7 @@ func main() {
 		}
 	}
 
+	defer git(o.candidatePath, []string{"checkout", jobSpec.Refs.Pulls[0].Ref})
```
why?

bbguimaraes replied:
It was annoying that the repository's `HEAD` ended up detached every time I executed it.

I haven't taken the time to figure out why yet.

bbguimaraes replied:
That happens because we use SHAs instead of refs when we checkout. From `git`'s man page:

```
       --detach
           Rather than checking out a branch to work on it, check out a commit
           for inspection and discardable experiments. This is the default
           behavior of "git checkout <commit>" when <commit> is not a branch
           name.
```

While this doesn't matter in the CI job, it is annoying when executed in a local repository, specially when iterating on a job — the HEAD is left pointing to the right commit, but detached. Was there any reason to choose one rather than the other? If there was, I think the current solution of `defer`ring a `git checkout` to restore the original HEAD is fine.

petr-muller replied:
> While this doesn't matter in the CI job, it is annoying when executed in a local repository, specially when iterating on a job — the HEAD is left pointing to the right commit, but detached. Was there any reason to choose one rather than the other? 

I think yes. My understanding is that when you run in a decorated Prowjob, you do not have `jobSpec.Refs.Pulls[0].Ref` checked out, you have it merged with the base. I'm not sure if the implementation does it in a way where the ref name is moved to the merge commit sha.

bbguimaraes replied:
Right, looking at the code, it is all done on top of the base ref.

I think `defer`ing a checkout is the simplest solution here. I have moved it inside the `if o.local` conditional, since that is not necessary in CI.

stevekuznetsov replied:
Creating named refs gets real messy real quick in CI -- you have to worry about collisions, etc. The CI knows exactly what it wants to be doing with SHAs and can do it more cleanly that way.

droslean reviewed (commented):
I like this because I found myself having a problem too when testing the tool locally.  The only thing that I don't like in this PR is that suddenly we do a lot of `git ` commands. I am starting to considering that maybe we need a library to handle all of that. Maybe `gopkg.in/libgit2/git2go.v25`?

bbguimaraes commented:
Base moved, will rebase.

bbguimaraes committed:
Add flag to derive JobSpec from local changes

bbguimaraes commented:
Rebased, added a flag to override the target namespace for ProwJobs (regular users cannot use the `ci` namespace, this allows them to `-n ci-stg`).

@petr-muller: you said you were rewriting the `git` calls. This touches them a bit, but it's a small change. I noticed I would have to change a few more things to avoid duplication after I rebased, but I can wait for your PR to merge to avoid unnecessary conflicts — maybe the changes then won't even be necessary.

bbguimaraes commented on cmd/pj-rehearse/main.go:
```diff
@@ -114,10 +119,44 @@ func main() {
+				Repo:    "release",
+				BaseSHA: baseSHA,
+				BaseRef: baseRef,
+				Pulls:   []pjapi.Pull{{SHA: pullSHA, Ref: pullRef}},
```
@stevekuznetsov: is anything horrible going to happen if we don't set the pull request number? We obviously have no value to set it to because these are local changes.

Nothing wrong happened in my tests with a PR number of `0`, but you surely know more of the dark corners of Prow.

stevekuznetsov replied:
I think that should be fine

droslean replied:
This should be empty only in a `periodic` job, right? Since that we rehearse only `presubmits` and also we populate the `jobspec` we won't have any problem with that.

bbguimaraes replied:
This is the fake `JobSpec` that is constructed based on local changes, where we don't have a pull request. The CI code path is unchanged.

petr-muller commented:
> you said you were rewriting the `git` calls. This touches them a bit, but it's a small change. I noticed I would have to change a few more things to avoid duplication after I rebased, but I can wait for your PR to merge to avoid unnecessary conflicts — maybe the changes then won't even be necessary.

:+1: my intention is to deduplicate the git checkout logic from the three `getX` methods, hopefully those are the parts where you spotted the duplication. Unfortunately it's not a stupid simple change, there are some quirks in error handling that need to be thought out well (some configs are important to load, some not that much).

stevekuznetsov commented on cmd/pj-rehearse/main.go:
```diff
@@ -61,8 +63,11 @@ func gatherOptions() options {
+	fs.BoolVar(&o.local, "local", false, "Whether this is a local execution or part of a CI job")
 	fs.BoolVar(&o.allowVolumes, "allow-volumes", false, "Allows jobs with extra volumes to be rehearsed")
 
+	fs.StringVar(&o.namespace, "n", "", "Short for --namespace")
```
Will anyone ever set this to something other than `ci-stg`? Why expose it to the user?

bbguimaraes replied:
Probably not (we do have some creative users), but I still think it is a nice option to have.

How about we keep the flag but default to `ci-stg` when `o.local`?

stevekuznetsov replied:
Creative users cause us headaches. Let's not expose this until we have a clear need, right now I think we simply

```go
namespace := "ci"
if o.runLocal {
    namespace = "ci-stg"
}
```

or whatever

stevekuznetsov commented on cmd/pj-rehearse/main.go:
```diff
@@ -114,10 +119,45 @@ func main() {
+			logrus.WithError(err).Error("could not get current branch")
+			gracefulExit(o.noFail, misconfigurationOutput)
+		}
+		baseRef, err := git(o.candidatePath, []string{"rev-parse", "--abbrev-ref", pullRef + "@{upstream}"})
```
I imagine people will not always have dev workflows that work with this. I think all of Prow is well-formed to use the SHA whenever present and fall back to the ref if not, so we _could_ set these to dummy values or at least be best-effort about them. Let's see if people end up erroring on this before we change anything.

bbguimaraes replied:
I went with the idea of an upstream branch because it is a concept that is native to git, but I am not crazy about it and am open to suggestions. Some others I considered:

- `master`: why not?
- a separate flag: sounds like too much work.
- some other git magic I am no aware of?

stevekuznetsov replied:
Let's use `master` if the user has not set an `upstream`? We don't have any other upstream branches on the `release` repo anyway right?

stevekuznetsov reviewed (approved):
These changes seem reasonable to me

bbguimaraes commented:
- rebased
- `master` is used when the branch doesn't have an upstream
- removed the `-n/--namespace` flag, `ci-stg` is used when `--local`
- the branch name (if it exists) is used to restore the original revision — it was easier and cleaner to implement it this way

stevekuznetsov reviewed (commented):
/lgtm
/approve
